src/MulticastClient.cs

```diff
@@ -0,0 +1,160 @@
+        {
+            this.multicastEndpoint = multicastEndpoint;
+
+            receiver = new UdpClient(multicastEndpoint.AddressFamily);
+
+            receiver.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+            receiver.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ExclusiveAddressUse, false);
+
+            receiver.Client.Bind(new IPEndPoint(IP6 ? IPAddress.IPv6Any : IPAddress.Any, multicastEndpoint.Port));
+
+            foreach (var address in nics.SelectMany(GetNetworkInterfaceLocalAddresses))
+            {
+                try
+                {
+                    receiver.Client.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.AddMembership, new MulticastOption(multicastEndpoint.Address, address));
+
+                    var sender = new UdpClient(multicastEndpoint.AddressFamily);
+
+                    sender.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+                    sender.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ExclusiveAddressUse, false);
+
+                    sender.Client.Bind(new IPEndPoint(address, multicastEndpoint.Port));
+
+                    sender.Client.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.AddMembership, new MulticastOption(multicastEndpoint.Address));
+                    sender.Client.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastLoopback, true);
+
+                    // Assigning multicastLoopbackAddress to first avalable address that we use for sending messages
+                    if (senders.TryAdd(address, sender) && multicastLoopbackAddress == null)
+                    {
+                        multicastLoopbackAddress = address;
+                    }
+                }
+                catch (SocketException ex) when (ex.SocketErrorCode == SocketError.AddressNotAvailable)
+                {
+                    // VPN NetworkInterfaces
+                }
+            }
+        }
+
+        public async Task SendAsync(byte[] message)
+        {
+            await Task.WhenAny(senders.Select(x => x.Value.SendAsync(message, message.Length, multicastEndpoint)));
+        }
+
+        public void Receive(Action<UdpReceiveResult> callback)
+        {
+            Task.Run(async () =>
+            {
+                try
+                {
+                    var task = receiver.ReceiveAsync();
+
+                    var ct1 = task.ContinueWith(x => Receive(callback), TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.RunContinuationsAsynchronously);
+
+                    var ct2 = task.ContinueWith(x => FilterMulticastLoopbackMessages(x.Result, callback), TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.RunContinuationsAsynchronously);
+
+                    await task.ConfigureAwait(false);
+                }
```

Comment:
This looks like magic to me.  Why `ContinueWith` and `ConfigureAwait`?  Is this what `await` is all about?

Reply:
I had to look this up as well (learned something new!) and found [this thread](https://stackoverflow.com/questions/27851073/why-would-i-bother-to-use-task-configureawaitcontinueoncapturedcontext-false) to be helpful.

In any case, I would suggest to write it as
```suggestion
await task.ConfigureAwait(continueOnCapturedContext: false);
```

Reply:
we have a receive task that we want to wait.
After the task is completed we immediately start listening socket again in new task task.ContinueWith(x => Receive(callback) ... - asynchronously (without awaiting for result)
Also, we processing result in a new task var ct2 = task.ContinueWith(x => FilterMulticastLoopbackMessages - asynchronously (without awaiting for result)
finally, await task.ConfigureAwait(false) - allows us safelly catch exception (ObjectDisposedException)

As we have more then 1 NIC multicast loop back we will receive as much as addresses were bond? we cont need this.
FilterMulticastLoopbackMessages - does allows us to receive loopback message only once.